src/silo/query_engine/query_plan.cpp

```diff
@@ -24,14 +24,24 @@ arrow::Result<QueryPlan> QueryPlan::makeQueryPlan(
    return query_plan;
 }
 
-arrow::Status QueryPlan::executeAndWriteImpl(std::ostream* output_stream) {
+arrow::Status QueryPlan::executeAndWriteImpl(
+   std::ostream* output_stream,
+   uint64_t timeout_in_seconds
+) {
    EVOBENCH_SCOPE("QueryPlan", "execute");
    SPDLOG_TRACE("{}", arrow_plan->ToString());
    arrow_plan->StartProducing();
    SPDLOG_TRACE("Plan started producing, will now read the resulting batches.");
    while (true) {
       arrow::Future<std::optional<arrow::ExecBatch>> future_batch = results_generator();
       SPDLOG_TRACE("await the next batch");
+      bool finished_batch_in_time = future_batch.Wait(timeout_in_seconds);
+      if (!finished_batch_in_time) {
+         return arrow::Status::ExecutionError(fmt::format(
```

Comment:
Based on some docs reading, you might want to call `StopProducing()` before returning: https://arrow.apache.org/docs/cpp/api/acero.html#_CPPv4N5arrow5acero8ExecPlan13StopProducingEv